server/repository/src/db_diesel/json_schema.rs

```diff
@@ -0,0 +1,101 @@
+use super::StorageConnection;
+
+use crate::RepositoryError;
+
+use diesel::prelude::*;
+use serde::{Deserialize, Serialize};
+
+table! {
+    json_schema (id) {
+        id -> Text,
+        schema -> Text,
+    }
+}
+
+#[derive(Clone, Queryable, Insertable, AsChangeset, Debug, PartialEq)]
+#[table_name = "json_schema"]
+pub struct JSONSchemaRow {
+    /// The json schema id
+    pub id: String,
+    /// Document path and name
+    pub schema: String,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct JSONSchema {
+    pub id: String,
+    pub schema: serde_json::Value,
+}
+
+pub struct JsonSchemaRepository<'a> {
+    connection: &'a StorageConnection,
+}
+
+fn schema_from_row(schema_row: JSONSchemaRow) -> Result<JSONSchema, RepositoryError> {
+    let parsed_schema: serde_json::Value =
+        serde_json::from_str(&schema_row.schema).map_err(|err| RepositoryError::DBError {
+            msg: "Invalid schema data".to_string(),
+            extra: format!("{}", err),
+        })?;
+    Ok(JSONSchema {
+        id: schema_row.id,
+        schema: parsed_schema,
+    })
+}
+
+fn row_from_schema(schema: &JSONSchema) -> Result<JSONSchemaRow, RepositoryError> {
+    let data = serde_json::to_string(&schema.schema).map_err(|err| RepositoryError::DBError {
+        msg: "Can't serialize data".to_string(),
+        extra: format!("{}", err),
+    })?;
+    Ok(JSONSchemaRow {
+        id: schema.id.to_owned(),
+        schema: data,
+    })
+}
+
+impl<'a> JsonSchemaRepository<'a> {
+    pub fn new(connection: &'a StorageConnection) -> Self {
+        JsonSchemaRepository { connection }
+    }
+
+    #[cfg(feature = "postgres")]
+    pub fn upsert_one(&self, schema: &JSONSchema) -> Result<(), RepositoryError> {
```

Comment:
i would assume json schema is immutable ?

Reply:
Yes, we still need it for sync though

Reply:
I hope we disable ever being able to edit the schema on central, otherwise we could have some nasty problems